### PR TITLE
RM-7426: Add class constraint to CreateSequence

### PR DESCRIPTION
**Pull request by dkudernatsch**

### CODE REVIEW

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 1]
```diff
@@ -161,7 +161,7 @@ static partial class EnumerableExtensions
 
     /// <summary>
     /// Generates a sequence of elements from the <paramref name="source"/> element by applying the specified next-element function, 
-    /// adding elements to the sequence while the current element satisfies the specified condition.
+    /// adding elements to the sequence while the current element satisfies the specified condition or the current element is not <see langword="null" />.
```
Comment (MichaelKetting) [inline SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:164]: we will need to add a breaking change ticket for this.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 2]
```diff
@@ -171,11 +171,12 @@ static partial class EnumerableExtensions
     /// A collection of elements containing the <paramref name="source"/> and all subsequent elements where each element satisfies a specified condition.
     /// </returns>
     public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source, Func<TSource, TSource> nextElementSelector, Func<TSource, bool> predicate)
+        where TSource : class
```
Comment (MichaelKetting) [inline SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:174]: That's also a breaking change ticket (separate from the first)

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 3]
```diff
@@ -222,12 +223,13 @@ public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source,
...
     {
-      return CreateSequenceWithCycleCheck (source, nextElementSelector, e => e != null, EqualityComparer<TSource>.Default, createCycleFoundException);
+      return CreateSequenceWithCycleCheck (source, nextElementSelector, e => true, EqualityComparer<TSource>.Default, createCycleFoundException);
     }
 
     /// <summary>
     /// Generates a sequence of elements from the <paramref name="source"/> element by applying the specified next-element function, 
-    /// adding elements to the sequence while the current element satisfies the specified condition. If a cycle is detected, an exception is thrown.
+    /// adding elements to the sequence while the current element satisfies the specified condition or the current element is not <see langword="null" />.
+    /// If a cycle is detected, an exception is thrown.
```
Comment (MichaelKetting) [inline SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:232]: Same here.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 4]
```diff
@@ -251,6 +253,7 @@ public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source,
         Func<TSource, bool> predicate,
         [CanBeNull] IEqualityComparer<TSource> equalityComparer,
         Func<TSource, TException> createCycleFoundException)
+        where TSource : class
```
Comment (MichaelKetting) [inline SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:256]: and 4th breaking change here.

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 5]
```diff
@@ -203,10 +203,10 @@ public void CreateSequence_WhilePredicateEvaluatesTrue_WithSingleElement ()
     }
 
     [Test]
-    public void CreateSequence_WhilePredicateEvaluatesTrue_WithValueType ()
+    public void CreateSequence_WhilePredicateEvaluatesTrue_StopsOnNull ()
     {
-      IEnumerable<int> actual = 4.CreateSequence (e => e - 1, e => e > 0);
-      Assert.That (actual.ToArray(), Is.EqualTo (new[] { 4, 3, 2, 1 }));
+      IEnumerable<string> actual = "start".CreateSequence (e => null, e => e != "end");
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:208]: hmm.. i think we can make an a bit more involved test.

Start-string: "ABCD"
Next Element:  value.SubString(0, len-1) and if string.length == 0 => null. 
The PRedicate can be "true"
Assert is then "ABCD", ABC, AB, A

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 6]
```diff
@@ -250,10 +250,10 @@ public void CreateSequenceWithCycleCheck_WhilePredicateEvaluatesTrue_WithSingleE
     }
 
     [Test]
-    public void CreateSequenceWithCycleCheck_WhilePredicateEvaluatesTrue_WithValueType ()
+    public void CreateSequenceWithCycleCheck_WhilePredicateEvaluatesTrue_StopsOnNull ()
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:253]: see previous note.

**Comment by iam-mholle** [comment]
Added to master with ef280408b8f1813f1e2d4e14a0ab1e0bfe232dc3.

**Pull request closed without merging**